### PR TITLE
Initial skeleton web app

### PR DESCRIPTION
**Pull request by bloodearnest**
- serves files index in json format
 - serves files contents

No authentication yet

### CODE REVIEW

File: hatch/models.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+from pydantic import BaseModel
+
+from hatch import config
+
+
+def get_sha(path):
+    """Calculate and cache the sha256 hash of a file.
+
+    We cache because a big workspace can have lots of large files, and serving
+    the index would involve calculated the hashes of all files, which could get
```
Comment (ghickman) [inline hatch/models.py:13]: s/calculated/calculating/

File: hatch/models.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    slow.
+
+    We cache it in a separate directory, but the same relative path. So the
+    cache for config.WORKSPACES/workspace1/output/file.txt is located at
+    config.CACHE/workspace1/output/file.txt.
+    """
+    sha_path = config.CACHE / path.relative_to(config.WORKSPACES)
+    # does cache file exist and is current?
+    if sha_path.exists() and sha_path.stat().st_mtime > path.stat().st_mtime:
```
Comment (ghickman) [inline hatch/models.py:22]: Could you break this out into a variable? I don't know the stat or st_mtime bits particularly well.  I think it's checking the cache file was modified more recently than the workspace file?  Would be good to make that more explicit though 🙂

File: hatch/app.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+app = FastAPI()
+
+
+@app.get("/workspace/{workspace}", response_model=FilesIndex)
+def workspace_index(workspace: str):
+    """Return a index of the files on disk in this workspace."""
+    path = config.WORKSPACES / workspace
+
+    # technically not async, but should be fast enought to
```
Comment (ghickman) [inline hatch/app.py:17]: s/enought/enough/

File: hatch/app.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+from hatch import config
+from hatch.models import FilesIndex
+
+
+app = FastAPI()
+
+
+@app.get("/workspace/{workspace}", response_model=FilesIndex)
+def workspace_index(workspace: str):
+    """Return a index of the files on disk in this workspace."""
```
Comment (ghickman) [inline hatch/app.py:14]: s/a index/an index/

**Review by ghickman — APPROVED** [review]
Bunch of small comment fixes, but otherwise this looks great.

Have you thought about enforcing static type checks since FastAPI is all typed, etc?

File: hatch/app.py (hunk 3 above)
Comment (bloodearnest) [reply hatch/app.py]: removed the comment, it was LIES

File: hatch/models.py (hunk 2 above)
Comment (bloodearnest) [reply hatch/models.py]: done

File: justfile [hunk 5]
```diff
@@ -16,7 +16,7 @@ test ARGS="":
     . scripts/setup_functions
     dev_setup
 
-    pytest --cov=. {{ ARGS }}
+    python -m pytest --cov=. --cov-report term-missing {{ ARGS }}
```
[the remark this answers is not in the document]
Comment (bloodearnest) [reply justfile:19]: I'll play with this - I'm planning to upstream some changes to justfile and other stuff back to the repo-template.

**Pull request merged**